### PR TITLE
UI Test Status Page

### PR DESCRIPTION
**Pull request by islemaster**
Introduce a `--with-status-page` option to runner.rb which generates a live-updating summary page for the test run for monitoring test progress.

screenshot from 2016-06-22 13-01-45

This page starts with a summary showing the branch name, commit under test and test run timestamp, and an overall test progress bar.

The rest of the page is a listing of all the tests by browser, displaying current status.  Each browser gets its own progress bar, too.

screenshot from 2016-06-22 13-02-02

The whole thing is backed by the test logs that we upload to S3, so this only works for test runs where we use the `--html` flag.  When a test completes (success _or_ failure now) we attach some metadata and push it to S3, where it will replace any previous test log for that branch-browser-feature combination (though the bucket is versioned, so the old logs are still available).

The test status page gets generated as a static file by runner.rb before any tests are run.  It is generated with test run metadata, including the branch name, commit, and the list of tests that are part of this test run.  Once opened, the page makes async requests to a new API `{origin}/test_status/{branch}/since/{timestamp}` to find out whether any tests have new output, and if so it hits `{origin}/test_status/{branch}/{test_name}` to retrieve results for a particular test.

I've hooked this up to CI so that on a DTT both the UI tests and the Eyes tests should put a link in Slack to their respective test status pages.

### CODE REVIEW

**Comment by islemaster** [comment]
PS: I know some of this is pretty hacky.  I just wanted to see if we could get something up and running quick, since it's an internal tool.

File: dashboard/app/controllers/test_status_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+require 'aws-sdk'
+require 'date'
+
+# API that serves UI test status based on uploaded test run S3 logs and their
+# metadata.  See runner.rb, test_status.haml and test_status.js for more
+# information.
+class TestStatusController < ApplicationController
```
Comment (aoby) [inline dashboard/app/controllers/test_status_controller.rb]: We might want to follow our [new API pattern](https://docs.google.com/document/d/10FKEASCjoANN2wLTIVSRM6EjPUqDj-VnVY1CM8LQll0) and put this under `/api/v1/`.

Comment (aoby) [inline dashboard/app/controllers/test_status_controller.rb]: nm I misread. I thought this was a REST API serving JSON only, but it looks like a regular controller backed by a view. Disregard.

Comment (islemaster) [inline dashboard/app/controllers/test_status_controller.rb]: :+1: I like that idea - I'd forgotten about that discussion before.  Will update.

File: dashboard/test/ui/runner.rb [hunk 2]
```diff
@@ -146,6 +148,9 @@ def upload_log_and_get_public_link(filename)
     f = `egrep -r "Given I am on .*#{scriptname.delete(' ').downcase}" . | cut -f1 -d ':' | sort | uniq | tr '\n' ,`
     $options.feature = f.split ','
   end
+  opts.on('--with-status-page', 'Generate a test status summary page for this test run') do
```
Comment (mehalshah) [inline dashboard/test/ui/runner.rb:151]: If this needs to be run with --html then you should force it here

**Comment by breville** [comment]
wow, this looks great.

**Comment by bcjordan** [comment]
LGTM. Excited to see this in action on next test run!

**Pull request merged**